File: tests/python/test_source_map.py
"""Tests for source map and active-k history propagation through Python bindings."""

import ucc


def test_hir_source_map_parallel_to_ops() -> None:
    """HirModule.source_map length matches num_ops."""
    hir = ucc.trace(ucc.parse("H 0\nT 0\nM 0"))
    assert len(hir.source_map) == hir.num_ops


def test_hir_source_map_contains_correct_lines() -> None:
    """T gate on line 2 maps to source_line 2."""
    hir = ucc.trace(ucc.parse("H 0\nT 0\nM 0"))
    # First HIR op is the T gate (H absorbed), should be line 2
    assert hir.source_map[0] == [2]


def test_compiled_source_map_parallel_to_bytecode() -> None:
    """CompiledModule source_map and active_k_history match bytecode length."""
    prog = ucc.lower(ucc.trace(ucc.parse("H 0\nT 0\nM 0")))
    assert len(prog.source_map) == prog.num_instructions
    assert len(prog.active_k_history) == prog.num_instructions


def test_active_k_history_shows_expansion_and_compaction() -> None:
    """k rises for T gate expansion and falls back after measurement."""
    prog = ucc.lower(ucc.trace(ucc.parse("H 0\nT 0\nM 0")))
    k_hist = list(prog.active_k_history)
    assert max(k_hist) >= 1, "T gate should expand k to at least 1"
    assert k_hist[-1] == 0, "Measurement should compact k back to 0"


def test_optimizer_preserves_source_map() -> None:
    """Peephole fusion keeps source_map in sync."""
    hir = ucc.trace(ucc.parse("H 0\nT 0\nT 0\nM 0"))
    pm = ucc.PassManager()
    pm.add(ucc.PeepholeFusionPass())
    pm.run(hir)
    assert len(hir.source_map) == hir.num_ops
    # The fused S gate should carry both T-gate source lines
    found = False
    for i, op in enumerate(hir):
        if op.as_dict()["op_type"] == "CLIFFORD_PHASE":
            lines = hir.source_map[i]
            assert 2 in lines and 3 in lines
            found = True
            break
    assert found, "Expected a fused CLIFFORD_PHASE op"


def test_empty_circuit_produces_empty_maps() -> None:
    """Empty circuit yields empty source_map and k_history."""
    prog = ucc.lower(ucc.trace(ucc.parse("")))
    assert len(prog.source_map) == 0
    assert len(prog.active_k_history) == 0

File: tests/test_source_map.cc
// Source map propagation tests
//
// Verifies that source line provenance threads correctly through the
// front-end and HIR optimization pipeline: parse -> trace -> optimize.

#include "clifft/circuit/parser.h"
#include "clifft/frontend/frontend.h"
#include "clifft/optimizer/hir_pass_manager.h"
#include "clifft/optimizer/peephole.h"

#include <catch2/catch_test_macros.hpp>
#include <cstdint>
#include <memory>
#include <vector>

using namespace clifft;

// Helper: full pipeline through trace (no optimizer)
static HirModule hir_from(const char* text) {
    return clifft::trace(clifft::parse(text));
}

// Helper: full pipeline through trace + peephole
static HirModule hir_optimized(const char* text) {
    auto hir = hir_from(text);
    HirPassManager pm;
    pm.add_pass(std::make_unique<PeepholeFusionPass>());
    pm.run(hir);
    return hir;
}

// =============================================================================
// Front-End source map population
// =============================================================================

TEST_CASE("Source map: trace populates source_map parallel to ops", "[source_map]") {
    auto hir = hir_from("H 0\nT 0\nM 0");
    REQUIRE(hir.source_map.size() == hir.ops.size());
    // T gate on line 2 produces at least one HIR op with source_line == 2
    bool found_t = false;
    for (size_t i = 0; i < hir.ops.size(); ++i) {
        if (hir.ops[i].op_type() == OpType::T_GATE) {
            REQUIRE(hir.source_map[i].size() == 1);
            REQUIRE(hir.source_map[i][0] == 2);
            found_t = true;
        }
    }
    REQUIRE(found_t);
}

TEST_CASE("Source map: Clifford-only circuit still has parallel source_map", "[source_map]") {
    auto hir = hir_from("H 0\nS 0\nM 0");
    REQUIRE(hir.source_map.size() == hir.ops.size());
    // Only ops should be MEASURE -- Cliffords are absorbed by the tableau
    for (size_t i = 0; i < hir.ops.size(); ++i) {
        REQUIRE(hir.source_map[i].size() == 1);
        REQUIRE(hir.source_map[i][0] > 0);
    }
}

TEST_CASE("Source map: multi-target gate emits one source_map entry per op", "[source_map]") {
    auto hir = hir_from("T 0 1 2");
    REQUIRE(hir.source_map.size() == hir.ops.size());
    // All T ops should trace back to line 1
    for (size_t i = 0; i < hir.ops.size(); ++i) {
        if (hir.ops[i].op_type() == OpType::T_GATE) {
            REQUIRE(hir.source_map[i] == std::vector<uint32_t>{1});
        }
    }
}

// =============================================================================
// Optimizer maintains source_map invariant
// =============================================================================

TEST_CASE("Source map: peephole fusion preserves parallel size", "[source_map]") {
    auto hir = hir_optimized("H 0\nT 0\nT 0");
    REQUIRE(hir.source_map.size() == hir.ops.size());
}

TEST_CASE("Source map: T plus T fusion deletes both source entries", "[source_map]") {
    // H 0 (line 1) -> absorbed by tableau
    // T 0 (line 2) + T 0 (line 3) -> fused to S, absorbed offline
    auto hir = hir_optimized("H 0\nT 0\nT 0");
    // S absorption eliminates both T-gate ops. No ops should remain.
    REQUIRE(hir.ops.empty());
    REQUIRE(hir.source_map.empty());
}

TEST_CASE("Source map: T plus T_dag cancellation removes both from map", "[source_map]") {
    // T 0 (line 1) + T_DAG 0 (line 2) -> cancelled entirely
    // M 0 (line 3) -> measurement remains
    auto hir = hir_optimized("T 0\nT_DAG 0\nM 0");
    REQUIRE(hir.source_map.size() == hir.ops.size());
    // No T_GATE should remain (CLIFFORD_PHASE no longer exists)
    for (size_t i = 0; i < hir.ops.size(); ++i) {
        REQUIRE(hir.ops[i].op_type() != OpType::T_GATE);
    }
}

TEST_CASE("Source map: absorbed Pauli residue removes only its source entries", "[source_map]") {
    auto hir = hir_optimized("R_Z(0.3) 0\nR_Z(0.7) 0\nR_X(0.2) 0");

    REQUIRE(hir.ops.size() == 1);
    REQUIRE(hir.ops[0].op_type() == OpType::PHASE_ROTATION);
    REQUIRE(hir.source_map == std::vector<std::vector<uint32_t>>{{3}});
}

TEST_CASE("Source map: terminal reset phase elimination removes only phase entries",
          "[source_map]") {
    auto hir = hir_optimized(
        "R 0 1\n"
        "H 0 1\n"
        "R_Z(0.3) 0 1\n"
        "X_ERROR(0.01) 0 1\n"
        "MR 0 1");

    REQUIRE(hir.source_map.size() == hir.ops.size());
    REQUIRE(hir.ops.size() == 10);

    size_t reset_entries = 0;
    size_t noise_entries = 0;
    size_t measure_reset_entries = 0;
    for (const auto& entry : hir.source_map) {
        REQUIRE(entry.size() == 1);
        REQUIRE(entry[0] != 3);
        reset_entries += entry[0] == 1;
        noise_entries += entry[0] == 4;
        measure_reset_entries += entry[0] == 5;
    }
    REQUIRE(reset_entries == 4);
    REQUIRE(noise_entries == 2);
    REQUIRE(measure_reset_entries == 4);
}
